### DISCUSSION TITLE
Need help with the config matcher of my middleware code

### DISCUSSION THREAD

**Discussion by AgusWork** [vercel/next.js#50740] [Help] [answered]
### Summary

I want to set unaccess a path to one of my pages, and i cant share the hole code because it is for work. But when i add the matches in my middleware nothing happens

import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';
import { getValidSubdomain } from '@/utils/subdomain';
import {  useRouter } from 'next/router';

// RegExp for public files
const PUBLIC_FILE = /\.(.*)$/; // Files

export async function middleware(req: NextRequest) {

  const router = useRouter();
  if (req.url.includes("/contact")) {
    console.log(req.url)
    return router.replace("/")

  }
  const regex = new RegExp('src/pages/contact')
  console.log(req.url)
  if (regex.test(req.url)) {
    return router.replace("/")
  }
  const url = req.nextUrl.clone();
  console.log(`Middleware - ${url}`);
  return NextResponse.redirect(new URL('/', req.url));

}
export const config = {
  matcher: [
    '/contact/:path*',
    '/pages/contact/',
    '/pages/contact',
    'pages/contact/:path*',
    'pages/contact/',
    '/pages/contact/:path*',
    '/contact/:path*',
    '/contact/',
    'contact/',
    'contact/:path*',
    'src/pages/contact/',
    'src/pages/contact',
    'src/pages/contact/:path*',
    '/src/pages/contact/',
    '/src/pages/contact',
    '/src/pages/contact/:path*',
  ],
};

### Additional information

```js
Y add some functions to stop going to that page but nothing happens, someone could help me?

The route of my archives are src > pages > contact > index.page.tsx 
Middleware is at the same level as pages. But i also try setting middleware in the root.
The src experimental directory is not available, i dont know if that could be the cause.
```

**Comment by devjiwonchoi** [comment, marked as answer]
The `matcher` for the `middleware` is not based on your project file path but on the `URL` path.

> `matcher` allows you to filter Middleware to run on specific paths.

In order to match the `/src/pages/contact/index.page.tsx` for example, the route path would be `/contact/index.page`, and the Matcher should be `/contact/:path`.

Since the route path `/contact` will be considered as `/contact/index`, the Matcher `/contact/:path` will match the `/contact` route path also.

> Note: For backward compatibility, Next.js always considers /public as /public/index. Therefore, a matcher of /public/:path will match.

If you want to match the pages under every `/contact` route, use `/contact/:path*` with an asterisk.

> Can have modifiers on named parameters (starting with :): /about/:path* matches /about/a/b/c because * is zero or more. ? is zero or one and + one or more

Note that the `middleware` file should be located at the root of your project.

> Use the file middleware.ts (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.

For more information, look [here](https://nextjs.org/docs/pages/building-your-application/routing/middleware#matcher).

Also, check out the [official guide](https://github.com/pillarjs/path-to-regexp#path-to-regexp-1) for the Matcher Path to Regexp.

# UPDATE

You can simply resolve this by setting the `as` to a `catch segment` you want the middleware to catch.

Example:

```js
import Link from 'next/link';
 
export default function RedirectFromContactToIndex() {
  return (
    <Link as="/middleware-catch-segment" href='/contact/index.page'>
      Contact
    </Link>
  );
}
```

```js
// middleware.js

import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';
 
export function middleware(request: NextRequest) {
  return NextResponse.redirect(new URL('/', request.url));
}
 
export const config = {
  matcher: '/middleware-catch-segment',
};
```

If you want the `middleware` to always redirect to `/` on `/contact/index.page`, you don't need to use `as` and just set the `matcher` to `/contact/index.page`.

Example:

```js
import Link from 'next/link';
 
export default function RedirectFromContactToIndex() {
  return (
    <Link href='/contact/index.page'>
      Contact
    </Link>
  );
}
```

```js
// middleware.js

import { NextResponse } from 'next/server';
import type { NextRequest } from 'next/server';
 
export function middleware(request: NextRequest) {
  return NextResponse.redirect(new URL('/', request.url));
}
 
export const config = {
  matcher: '/contact/index.page',
};
```

**Reply by AgusWork** [reply]
Yes its the first try that i've use but the middleware is still going to that page.

**Comment by devjiwonchoi** [comment]
@AgusWork 
I found out that the file path `/src/page/contact/index.page.tsx` as route path is `/contact/index.page`. Updated the answer and it should work.

If there isn't a specific reason for the name `index.page.tsx`, I would recommend you to change it to `index.tsx` so that the route can stay simple as `/contact`.

**Reply by AgusWork** [reply]
Thank you so much for the answer i will try with that path

**Reply by AgusWork** [reply]
I dont know why it has that name is for a work that is was already display like that when i joined

**Reply by devjiwonchoi** [reply]
@AgusWork 
My bad.. I left a wrong answer. I've updated my answer please check it out.

**Reply by devjiwonchoi** [reply]
@AgusWork 
I've opened a [repository](https://github.com/devjiwonchoi/next.js-page-router-example) to test your case. Check it out for a clearer example.

**Reply by AgusWork** [reply]
I check it out with the changes that you send me, but it has no change, maibe is the form of how the files are disposed, or the way that is this program made. I will search for another way to unaccess this page with server side rendering.

**Reply by devjiwonchoi** [reply]
@AgusWork  Does it not work with the new update?

```js
<Link as="/middleware-catch-segment" href='/contact/index.page'>
  Contact
</Link>
```

```js
export function middleware(request: NextRequest) {
  return NextResponse.redirect(new URL('/', request.url));
}
 
export const config = {
  matcher: '/middleware-catch-segment',
};
```

Also, note that the `middleware` should be right below `src` folder.
`/src/middleware.ts`

This should definitely work could you please try it out again?

**Reply by AgusWork** [reply]
It doesnt, i dont know why, but this is the next config /** @type {import('next').NextConfig} */
const nextConfig = {
reactStrictMode: true,
pageExtensions: ['page.tsx', 'page.ts', 'page.jsx', 'page.js'],
env: {
API_URL: process.env.API_URL,
},
};

module.exports = nextConfig;

It doesnt have the src directory available, the middleware component it has to be in the root or below the src?

**Reply by devjiwonchoi** [reply]
What is the version of your `next`?

For this,

> pageExtensions: ['page.tsx', 'page.ts', 'page.jsx', 'page.js'],

the `index.page.tsx` would be recognized as `index.tsx`, no needed to do something like `matcher: '/contact/index.page',` just `matcher: '/contact'`.

Also, the middleware SHOULD be right below the `src` if your `pages` are inside the `src`.

> Use the file middleware.ts (or .js) in the root of your project to define Middleware. For example, at the same level as pages or app, or inside src if applicable.

**Reply by AgusWork** [reply]
My version of next is 13.2.4

And this is the disposal of the files

![image](https://github.com/vercel/next.js/assets/100211046/542091a1-98d7-44c9-877f-af4b1435a832)

**Reply by devjiwonchoi** [reply]
One thing I want to make clear, are you trying to make the route `/contact` always redirect to `/`?
For example, `http://localhost:3000/contact` to `http://localhost:3000`?

**Reply by AgusWork** [reply]
yes, sorry for the last response, i had to relax
